### PR TITLE
feat(utilities): add `objectKeys` function

### LINKED ISSUE #457: request: add `objectKeys` (a strictly typed `Object.keys`) to `@sapphire/utilities`
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

The types built into TypeScript for `Object.keys` intentionally aren't fully strictly typed, however there are instances where a strictly typed `Object.keys` can be useful. This is where a custom `objectKeys` can come in.

### Desired solution

Add a function like this and export it:

```ts
export function objectKeys<T extends NonNullObject>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[];
}
```

Then implement unit tests for this

### Alternatives considered

N.A.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by RealShadowNova**
Resolves #457

### CODE REVIEW

File: packages/utilities/tests/objectKeys.test.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { objectKeys } from '../src';
+
+describe('objectKeys', () => {
+	test('GIVEN basic THEN returns expected', () => {
+		const source = { a: 'Hello', b: 420 };
+		const expected = ['a', 'b'];
+
+		expect(objectKeys(source)).toEqual(expected);
```
Comment (kyranet) [inline packages/utilities/tests/objectKeys.test.ts:8]: Same as the other PR, please assert the input type + test readonly objects.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/utilities/src/lib/objectKeys.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import type { NonNullObject } from './utilityTypes';
+
+export function objectKeys<T extends NonNullObject>(obj: T): (keyof T)[] {
+	return Object.keys(obj) as (keyof T)[];
```
Comment (imranbarbhuiya) [inline packages/utilities/src/lib/objectKeys.ts:4]:
```suggestion
export function objectKeys<T extends NonNullObject>(obj: T):  T extends ArrayLike<any> ? number[] : (keyof T)[] {
	return Object.keys(obj) as T extends ArrayLike<any> ? number[] : (keyof T)[]
```

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

**Comment by RealShadowNova** [comment]
Should I change the return type to a `string` or to a template literal containing `number` in the case of an array?
image

**Comment by imranbarbhuiya** [comment]
template literal containing `number`

**Review by favna — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

File: packages/utilities/src/lib/objectKeys.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import type { NonNullObject } from './utilityTypes';
+
+export function objectKeys<T extends NonNullObject>(obj: T): T extends ArrayLike<any> ? `${number}`[] : (keyof T)[] {
+	return Object.keys(obj) as T extends ArrayLike<any> ? `${number}`[] : (keyof T)[];
```
Comment (favna) [inline packages/utilities/src/lib/objectKeys.ts:4]:
```suggestion
	return Object.keys(obj) as T extends ArrayLike<infer Values> ? Values[] : (keyof T)[];
```

**Review by favna — CHANGES_REQUESTED** [review]

**Review by favna — APPROVED** [review]

**Pull request merged**